`new-e2e-containers-eks` flakiness: requeue unacked language detection events

rdesgroppes opened the pull request:
### What does this PR do?
This is a tentative fix for a race condition in language detection where events appear to be silently dropped when the Cluster Agent is not the leader (rolling updates, multiple replicas...), which may be causing `TestAdmissionControllerWithAutoDetectedLanguage` to flake.

The change proposes to make the non-leader case to return `ErrNotLeader` instead of `nil`.
This should allow the workqueue to requeue the event with exponential backoff until a leader is elected and can process it.

### Motivation
See logs (`main` pipeline):
- https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/1279349200#L1238
- https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/1288699301#L286
```
=== FAIL: test/new-e2e/tests/containers TestEKSSuite/TestAdmissionControllerWithAutoDetectedLanguage (364.58s)
    base_test.go:37: START  eksSuite/TestAdmissionControllerWithAutoDetectedLanguage 2025-12-09 15:05:42.472157044 +0000 UTC m=+329.810291814
    suite.go:384: No change in provisioners, skipping environment update
    k8s_test.go:1268: 
        	Error Trace:	/go/src/github.com/DataDog/datadog-agent/test/new-e2e/tests/containers/k8s_test.go:1287
        	            				/usr/local/go/src/runtime/asm_amd64.s:1700
        	Error:      	Should be true
    k8s_test.go:1268: 
        	Error Trace:	/go/src/github.com/DataDog/datadog-agent/test/new-e2e/tests/containers/k8s_test.go:1268
        	            				/go/src/github.com/DataDog/datadog-agent/test/new-e2e/tests/containers/k8s_test.go:1228
        	            				/go/pkg/mod/github.com/stretchr/testify@v1.11.1/suite/suite.go:196
        	            				/go/pkg/mod/github.com/!data!dog/dd-trace-go/v2@v2.2.2/internal/civisibility/integrations/gotesting/instrumentation_orchestrion.go:242
        	Error:      	Condition never satisfied
        	Test:       	TestEKSSuite/TestAdmissionControllerWithAutoDetectedLanguage
        	Messages:   	The deployment with name mutated-with-auto-detected-language in namespace workload-mutated-lib-injection does not exist or does not have the auto detected languages annotation
```

The test `TestAdmissionControllerWithAutoDetectedLanguage` has been flaky since ~Dec 2024.
The test waits up to 5 minutes for the Cluster Agent to add the `internal.dd.datadoghq.com/.*\.detected_langs` annotation on deployments, but this annotation sometimes never appears.

Prior art:
- #24732
- #32422
- #32469
- #40526

From `pkg/clusteragent/languagedetection/patcher.go`:
```go
if !lp.isLeader() {
    lp.logger.Tracef("Skipping patch for %s: %s/%s - not leader", ...)
    return nil
    //     ^^^ event is dropped, not requeued
}
```

Since all Cluster Agent pods receive events independently from `workloadmeta`, and the workqueue handles retries when errors are returned, returning `nil` causes the event to be forgotten rather than retried.

### Describe how you validated your changes
Added `TestPatcherRequeuesWhenNotLeader` to verify events are requeued when not leader.

### Additional Notes
:warning: This is all new to me, and it's very possible that I may have missed something.

Another culprit might be involved (e.g., the language detection itself may be timing out or failing).

rdesgroppes committed:
`new-e2e-containers-eks` flakiness: requeue language detection events

### What does this PR do?
This is a tentative fix for a race condition in language detection where events appear to be silently dropped when the Cluster Agent is not the leader (rolling updates, multiple replicas...), which may be causing `TestAdmissionControllerWithAutoDetectedLanguage` to flake.

The change proposes to changes the non-leader case to return `apiserver.ErrNotLeader` instead of `nil`.
This should allow the workqueue to requeue the event with exponential backoff until a leader is elected and can process it.

### Motivation
The test `TestAdmissionControllerWithAutoDetectedLanguage` has been flaky since ~Dec 2024.
The test waits up to 5 minutes for the Cluster Agent to add the `internal.dd.datadoghq.com/.*\.detected_langs` annotation on deployments, but this annotation sometimes never appears.

Prior art:
- #32422
- #32469
- #40526

From pkg/clusteragent/languagedetection/patcher.go:251-255:
```go
if !lp.isLeader() {
    lp.logger.Tracef("Skipping patch for %s: %s/%s - not leader", ...)
    return nil
    //     ^^^ event is dropped, not requeued
}
```

Since all Cluster Agent pods receive events independently from `workloadmeta`, and the workqueue handles retries when errors are returned, returning `nil` causes the event to be forgotten rather than retried.

### Describe how you validated your changes
Added `TestPatcherRequeuesWhenNotLeader` to verify events are requeued when not leader.

### Additional Notes
This is all new to me, and it's very possible that I may have missed something.

Another culprit might be involved (e.g., the language detection itself may be timing out or failing).

rdesgroppes commented:
something

rdesgroppes commented:
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

chouetz reviewed (commented):
SGTM for agent-devx, nice catch for the flakiness detection!

rdesgroppes commented:
/merge

rdesgroppes committed:
Set shorter backoff as advised by Minyi Zhu

rdesgroppes commented:
/merge

rdesgroppes committed:
Push events to all Cluster Agent pods

Problem: language detection events were only pushed to the leader's
local workloadmeta. If leadership changed between receiving the HTTP
request and processing the patch, the event would be stuck in the old
leader's queue, while the new leader never received it. This causes the
annotation to never be created and tests to fail.

Proposed solution: move the `mergeAndFlush` logic from `leaderHandler`
to `preHandler` so that:
- all pods (leader _and followers_) push events to their local
  workloadmeta,
- events survive leadership changes: the new leader already has them in
  its queue,
- the patcher still checks `isLeader` before patching, so only the
  leader patches,
- the HTTP body is restored after processing for downstream handlers
  (`Forward` or `leaderHandler`).
